core/base/index_range.hpp

```diff
@@ -0,0 +1,238 @@
+    /** Pre-decrement. */
+    constexpr index_iterator& operator--() { return *this -= 1; }
+
+    /** Post-decrement. */
+    constexpr index_iterator operator--(int)
+    {
+        auto tmp{*this};
+        operator--();
+        return tmp;
+    }
+
+    // RandomAccessIterator requirements
+    /** Iterator advance. */
+    constexpr friend index_iterator operator+(index_iterator a,
+                                              difference_type n)
+    {
+        return a += n;
+    }
+
+    /** reverse advance */
+    constexpr friend index_iterator operator+(difference_type n,
+                                              index_iterator a)
+    {
+        return a + n;
+    }
+
+    /** Iterator backwards advance. */
+    constexpr index_iterator& operator-=(difference_type n)
+    {
+        return *this += -n;
+    }
+
+    /** Iterator backwards advance. */
+    constexpr friend index_iterator operator-(index_iterator a,
+                                              difference_type n)
+    {
+        return a + -n;
+    }
+
+    /** Subscript. */
+    constexpr reference operator[](difference_type n) const
+    {
+        return *(*this + n);
+    }
+
+    /** less than */
+    constexpr friend bool operator<(index_iterator a, index_iterator b)
+    {
+        return b - a > 0;
+    }
+
+    /** greater than */
+    constexpr friend bool operator>(index_iterator a, index_iterator b)
+    {
+        return b < a;
+    }
+
+    /** greater equal */
+    constexpr friend bool operator>=(index_iterator a, index_iterator b)
+    {
+        return !(a < b);
+    }
+
+    /** less equal */
+    constexpr friend bool operator<=(index_iterator a, index_iterator b)
+    {
+        return !(a > b);
+    }
+
+private:
+    index_type idx_;
+};
+
+
+/**
+ * Represents an index range that will be iterated through in order.
+ * The following two loops are equivalent (assuming begin <= end):
+ * ```cpp
+ * for (int i = begin; i < end; i++);
+ * for (auto i : irange<int>{begin, end});
+ * ```
+ *
+ * @tparam IndexType  the type of the indices.
+ */
+template <typename IndexType>
+class irange : public iterator_range<index_iterator<IndexType>> {
+public:
+    using index_type = IndexType;
+    using iterator = index_iterator<index_type>;
+
+    /**
+     * Constructs an index range consisting of the values in [begin, end).
+     *
+     * @param begin  the first index in the range
+     * @param end  one past the last index in the range. It must not be smaller
+     *             than `begin`.
+     */
+    constexpr explicit irange(index_type begin, index_type end)
+        : iterator_range<iterator>{iterator{begin}, iterator{end}}
+    {
+        assert(begin <= end);
+    }
+
+    /**
+     * Constructs an index range consisting of the values in [0, end).
+     *
+     * @param end  one past the last index in the range. It must not be a
+     *             negative number.
+     */
+    constexpr explicit irange(index_type end) : irange{0, end} {}
```

Comment:
Does this have to be `explicit` ? You would need to have `static_cast`s in most places as the `size()`, in many places might be `unsigned` ? Maybe you can already move the `static_cast` inside here ? 
```suggestion
    constexpr irange(index_type end) : irange{0, static_cast<index_type>(end)} {}
```

Reply:
`explicit` disables implicit conversions _to_ `irange`, but it doesn't involve the parameter types. The need for `static_cast` comes from the fact that `std::size_t` is unsigned and either the use of `{}` for construction or the want to avoid signed-unsigned conversion warnings.